test/leakcheck/leakcheck.go

```diff
@@ -0,0 +1,94 @@
+/*
+ *
+ * Copyright 2017 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+// Package leakcheck contains functions to check leaked goroutines.
+//
+// Call "defer leakcheck.Check(t)()" at the beginning of tests.
+package leakcheck
+
+import (
+	"runtime"
+	"sort"
+	"strings"
+	"testing"
+	"time"
+)
+
+// interestingGoroutines returns all goroutines we care about for the purpose of
+// leak checking. It excludes testing or runtime ones.
+func interestingGoroutines() (gs []string) {
+	buf := make([]byte, 2<<20)
+	buf = buf[:runtime.Stack(buf, true)]
+	for _, g := range strings.Split(string(buf), "\n\n") {
+		sl := strings.SplitN(g, "\n", 2)
+		if len(sl) != 2 {
+			continue
+		}
+		stack := strings.TrimSpace(sl[1])
+		if strings.HasPrefix(stack, "testing.RunTests") {
+			continue
+		}
+
+		if stack == "" ||
+			strings.Contains(stack, "testing.Main(") ||
+			strings.Contains(stack, "testing.tRunner(") ||
+			strings.Contains(stack, "testing.(*M).") ||
+			strings.Contains(stack, "runtime.goexit") ||
+			strings.Contains(stack, "created by runtime.gc") ||
+			strings.Contains(stack, "created by runtime/trace.Start") ||
+			strings.Contains(stack, "created by google3/base/go/log.init") ||
+			strings.Contains(stack, "interestingGoroutines") ||
+			strings.Contains(stack, "runtime.MHeap_Scavenger") ||
+			strings.Contains(stack, "signal.signal_recv") ||
+			strings.Contains(stack, "sigterm.handler") ||
+			strings.Contains(stack, "runtime_mcall") ||
+			strings.Contains(stack, "(*loggingT).flushDaemon") ||
+			strings.Contains(stack, "goroutine in C code") {
+			continue
+		}
+		gs = append(gs, g)
+	}
+	sort.Strings(gs)
+	return
+}
+
+// Check snapshots the currently-running goroutines and returns a function to be
+// run at the end of tests to see whether any goroutines leaked.
+func Check(t testing.TB) {
+	// Loop, waiting for goroutines to shut down.
+	// Wait up to 10 seconds, but finish as quickly as possible.
+	deadline := time.Now().Add(10 * time.Second)
+	for {
+		var leaked []string
+		for _, g := range interestingGoroutines() {
+			leaked = append(leaked, g)
+		}
+		if len(leaked) == 0 {
+			return
+		}
+		if time.Now().Before(deadline) {
+			time.Sleep(50 * time.Millisecond)
+			continue
+		}
+		for _, g := range leaked {
+			t.Errorf("Leaked goroutine: %v", g)
+		}
+		return
+	}
+
+}
```

Comment:
Should we add a leakcheck_test.go to make sure it works?
```go
func TestCheck(t *testing.T) {
  go func() { time.Sleep(5) }()
  if ig := interestingGoroutines(); len(ig) == 0 {
    t.Error()
  }
  Check()
}
```

Reply:
Done